modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java

```diff
@@ -604,10 +611,71 @@ public RelDataType deriveType(SqlValidatorScope scope, SqlNode expr) {
                 });
     }
 
+    /** Check string contain only digit characters.
+     *
+     * @param strVal String representation.
+     * @param pos Current validator position.
+     *
+     * @throws CalciteException if unexpected input string found.
+     **/
+    static void checkStringContainDigitsOnly(@Nullable String strVal, SqlParserPos pos) throws CalciteException {
+        if (strVal == null) {
+            return;
+        }
+
+        Matcher matcher = DIGITS_REG.matcher(strVal);
+        boolean ret = matcher.find();
+
+        if (!ret) {
+            var ex = RESOURCE.invalidCharacterForCast(strVal);
+            throw SqlUtil.newContextException(pos, ex);
+        }
+    }
+
     /** Check appropriate type cast availability. */
     private void checkTypesInteroperability(SqlValidatorScope scope, SqlNode expr) {
         boolean castOp = expr.getKind() == SqlKind.CAST;
 
+        if (expr.getKind() == SqlKind.IN || expr.getKind() == SqlKind.NOT_IN) {
```

Comment:
It would be better to have some comments explaining what problem this code solves.
This can be moved into a separated method.